File: slave/receive.py
# -*- coding: utf-8 -*-
import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import serial
import threading
import time
import datetime

PORT = '/dev/tty.usbserial-5AB31140701'  # SLAVE1 端口
BAUD = 921600


class SlaveTerminal:
    def __init__(self, port: str, baud: int):
        self.port = port
        self.baud = baud
        self.ser = None
        self.running = False
        self.rx_thread = None
    
    def start(self):
        self.ser = serial.Serial(self.port, self.baud, timeout=0.1)
        self.ser.reset_input_buffer()
        self.running = True
        
        self.rx_thread = threading.Thread(target=self._receive_loop, daemon=True)
        self.rx_thread.start()
        
        print(f'SLAVE 终端已启动 ({self.port}, {self.baud}bps)')
        print('等待 MASTER 数据...\n')
        self._print_help()
    
    def stop(self):
        self.running = False
        if self.rx_thread:
            self.rx_thread.join(timeout=1)
        if self.ser:
            self.ser.close()
        print('终端已停止')
    
    def _print_help(self):
        print('命令:')
        print('  <消息>        向 MASTER 发送消息')
        print('  /h            显示帮助')
        print('  /q            退出\n')
    
    def _receive_loop(self):
        while self.running:
            try:
                data = self.ser.read(4096)
                if data:
                    self._handle_receive(data)
            except Exception as e:
                if self.running:
                    print(f'[接收错误] {e}')
    
    def _handle_receive(self, data: bytes):
        now = datetime.datetime.now().strftime('%H:%M:%S.%f')[:-3]
        try:
            text = data.decode('utf-8')
            print(f'\r[{now}] ← MASTER: "{text}"')
        except UnicodeDecodeError:
            print(f'\r[{now}] ← MASTER (HEX): {data.hex(" ").upper()}')
        print('> ', end='', flush=True)
    
    def send(self, data: bytes):
        self.ser.write(data)
        now = datetime.datetime.now().strftime('%H:%M:%S.%f')[:-3]
        print(f'[{now}] → MASTER: {data!r} ({len(data)}B)')
    
    def run_interactive(self):
        self.start()
        try:
            while self.running:
                try:
                    line = input('> ').strip()
                    if not line:
                        continue
                    
                    if line == '/q':
                        break
                    elif line == '/h':
                        self._print_help()
                    elif line.startswith('/'):
                        print(f'未知命令: {line}')
                        self._print_help()
                    else:
                        self.send(line.encode('utf-8'))
                except EOFError:
                    break
        except KeyboardInterrupt:
            pass
        finally:
            self.stop()
    
    def run_receive_only(self):
        self.start()
        print('纯接收模式，按 Ctrl+C 退出\n')
        try:
            while self.running:
                time.sleep(0.1)
        except KeyboardInterrupt:
            pass
        self.stop()


def main():
    terminal = SlaveTerminal(PORT, BAUD)
    
    if len(sys.argv) < 2:
        terminal.run_interactive()
    elif sys.argv[1] == '-r':
        terminal.run_receive_only()
    elif sys.argv[1] == '-h':
        print('用法:')
        print('  python slave/receive.py        交互模式')
        print('  python slave/receive.py -r     纯接收模式')
        print('  python slave/receive.py -h     显示帮助')
    else:
        msg = ' '.join(sys.argv[1:])
        terminal.start()
        terminal.send(msg.encode())
        time.sleep(0.1)
        terminal.stop()


if __name__ == '__main__':
    main()

File: nodejs/slave/receive.mjs
import { SerialPort } from 'serialport';
import readline from 'readline';

const PORT = '/dev/cu.wchusbserial5AB50010561';
const BAUD = 921600;

class SlaveTerminal {
  constructor(port, baud) {
    this.port = port;
    this.baud = baud;
    this.serialPort = null;
    this.running = false;
    this.rl = null;
  }

  async start() {
    this.serialPort = new SerialPort({
      path: this.port,
      baudRate: this.baud,
      dataBits: 8,
      parity: 'none',
      stopBits: 1
    });

    await new Promise((resolve, reject) => {
      this.serialPort.on('open', resolve);
      this.serialPort.on('error', reject);
    });

    this.serialPort.on('data', (data) => this.handleReceive(data));
    this.serialPort.on('error', (err) => {
      if (this.running) console.log(`\n[接收错误] ${err.message}`);
    });

    this.running = true;

    console.log(`SLAVE 终端已启动 (${this.port}, ${this.baud}bps)`);
    console.log('等待 MASTER 数据...\n');
    this.printHelp();
  }

  stop() {
    this.running = false;
    if (this.rl) {
      this.rl.close();
    }
    if (this.serialPort && this.serialPort.isOpen) {
      this.serialPort.close();
    }
    console.log('终端已停止');
  }

  printHelp() {
    console.log('命令:');
    console.log('  <消息>        向 MASTER 发送消息');
    console.log('  /h            显示帮助');
    console.log('  /q            退出\n');
  }

  getTimestamp() {
    const now = new Date();
    return now.toTimeString().split(' ')[0] + '.' + now.getMilliseconds().toString().padStart(3, '0');
  }

  handleReceive(data) {
    const timestamp = this.getTimestamp();
    try {
      const text = data.toString('utf8');
      process.stdout.write(`\r[${timestamp}] ← MASTER: "${text}"\n> `);
    } catch {
      const hexStr = data.toString('hex').toUpperCase().match(/.{2}/g).join(' ');
      process.stdout.write(`\r[${timestamp}] ← MASTER (HEX): ${hexStr}\n> `);
    }
  }

  send(data) {
    this.serialPort.write(data);
    const timestamp = this.getTimestamp();
    console.log(`[${timestamp}] → MASTER: ${JSON.stringify(data.toString())} (${data.length}B)`);
  }

  async runInteractive() {
    await this.start();

    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const prompt = () => {
      this.rl.question('> ', (line) => {
        if (!this.running) return;

        const trimmed = line.trim();
        if (!trimmed) {
          prompt();
          return;
        }

        if (trimmed === '/q') {
          this.stop();
          return;
        }

        if (trimmed === '/h') {
          this.printHelp();
          prompt();
          return;
        }

        if (trimmed.startsWith('/')) {
          console.log(`未知命令: ${trimmed}`);
          this.printHelp();
          prompt();
          return;
        }

        this.send(Buffer.from(trimmed, 'utf8'));
        prompt();
      });
    };

    prompt();
  }

  async runReceiveOnly() {
    await this.start();
    console.log('纯接收模式，按 Ctrl+C 退出\n');

    await new Promise(resolve => {
      process.on('SIGINT', resolve);
    });

    this.stop();
  }
}

function printUsage() {
  console.log('用法:');
  console.log('  node slave/receive.js        交互模式');
  console.log('  node slave/receive.js -r     纯接收模式');
  console.log('  node slave/receive.js -h     显示帮助');
}

async function main() {
  const terminal = new SlaveTerminal(PORT, BAUD);
  const args = process.argv.slice(2);

  try {
    if (args.length === 0) {
      await terminal.runInteractive();
    } else if (args[0] === '-r') {
      await terminal.runReceiveOnly();
    } else if (args[0] === '-h') {
      printUsage();
    } else {
      const msg = args.join(' ');
      await terminal.start();
      terminal.send(Buffer.from(msg, 'utf8'));
      await new Promise(resolve => setTimeout(resolve, 100));
      terminal.stop();
    }
  } catch (err) {
    console.error(`错误: ${err.message}`);
    terminal.stop();
    process.exit(1);
  }
}

main();
